File: app/src/main/java/org/attendr/helpers/views/PercentageCircle.java
package org.attendr.helpers.views;

import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;
import android.os.Handler;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.AccelerateInterpolator;

import org.attendr.R;

public class PercentageCircle extends View {

    private ValueAnimator animator;
    private Paint paint;
    private Paint paintGray;
    private Paint paintText;
    private Paint centerPaint;
    private RectF bounds;
    private RectF boundsGray;
    private RectF centerBounds;

    private int width;
    private int height;

    private int angle = 1;
    private String text = "";

    private int strokeWidth;

    public PercentageCircle(Context context) {
        super(context);
        init();
    }

    public PercentageCircle(Context context, AttributeSet attrs) {
        super(context, attrs);
        TypedArray attributes = context.getTheme().obtainStyledAttributes(attrs, R.styleable.PercentageCircle, 0, 0);
        strokeWidth = attributes.getDimensionPixelSize(R.styleable.PercentageCircle_stroke, 3);
        attributes.recycle();
        init();
    }

    public PercentageCircle(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        TypedArray attributes = context.getTheme().obtainStyledAttributes(attrs, R.styleable.PercentageCircle, 0, 0);
        strokeWidth = attributes.getDimensionPixelSize(R.styleable.PercentageCircle_stroke, 3);
        attributes.recycle();
        init();
    }

    private void init() {
        paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        paint.setStyle(Paint.Style.FILL);
        paint.setColor(ContextCompat.getColor(getContext(), R.color.colorVividBlue));

        paintGray = new Paint(Paint.ANTI_ALIAS_FLAG);
        paint.setStyle(Paint.Style.FILL);
        paintGray.setColor(ContextCompat.getColor(getContext(), R.color.colorLightGray));

        centerPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        centerPaint.setStyle(Paint.Style.FILL);
        centerPaint.setColor(ContextCompat.getColor(getContext(), android.R.color.white));

        paintText = new Paint(Paint.ANTI_ALIAS_FLAG);
        paintText.setStyle(Paint.Style.FILL);
        paintText.setColor(ContextCompat.getColor(getContext(), R.color.colorVividBlue));

        recalcSize();
    }

    public void recalcSize() {
        int heightOrWidth = Math.min(width, height);
        strokeWidth = heightOrWidth / 15;
        int diameter = heightOrWidth;
        int top = strokeWidth / 2;
        int bottom = diameter - strokeWidth / 2;
        paint.setStrokeWidth(strokeWidth);
        bounds = new RectF(top, top, bottom, bottom);
        boundsGray = new RectF(top, top, bottom, bottom);
        int newTop = top + strokeWidth;
        int newBottom = bottom - strokeWidth;
        centerBounds = new RectF(newTop, newTop, newBottom, newBottom);
        paintText.setTextSize(heightOrWidth / 4);
        invalidate();
    }

    public int getProgress() {
        return angle;
    }

    public void setProgress(int percentage) {
        int angle = (int) (360 * (percentage / 100.0f));
        if (0 <= percentage && percentage <= 100) {
            text = Integer.toString(percentage) + "%";
            if (animator != null) {
                animator.cancel();
                animator.setIntValues(getProgress(), angle);
            } else {
                animator = ValueAnimator.ofInt(getProgress(), angle);
                animator.setInterpolator(new AccelerateInterpolator());
                animator.setDuration(800);
                animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
                    @Override
                    public void onAnimationUpdate(ValueAnimator animation) {
                        PercentageCircle.this.angle = (Integer) animation.getAnimatedValue();
                        invalidate();
                    }
                });
            }
            animator.start();
        }
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        width = getMeasuredWidth();
        height = getMeasuredHeight();
        setMeasuredDimension(widthMeasureSpec, heightMeasureSpec);
        recalcSize();
        invalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        canvas.drawArc(boundsGray, 0, 360, true, paintGray);
        canvas.drawArc(bounds, -91, angle, true, paint);
        canvas.drawArc(centerBounds, 0, 360, true, centerPaint);
        int xPos = (canvas.getWidth() / 2);
        int yPos = xPos - (int) ((paintText.descent() + paintText.ascent()) / 2);
        canvas.drawText(text, xPos - (int) paintText.measureText(text) / 2, yPos, paintText);
    }
}

File: app/src/main/java/org/attendr/helpers/views/PercentageCircle.kt
package org.attendr.helpers.views

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import android.support.v4.content.ContextCompat
import android.util.AttributeSet
import android.view.View
import android.view.animation.AccelerateInterpolator
import org.attendr.R

/**
 * @author Pauldg7@gmail.com (Paul Gillis)
 */
class PercentageCircle(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : View(context, attrs, defStyleAttr) {

    private var animator: ValueAnimator? = null

    private var paint: Paint
    private var paintGray: Paint
    private var paintText: Paint
    private var centerPaint: Paint

    private var bounds: RectF? = null
    private var boundsGray: RectF? = null
    private var centerBounds: RectF? = null

    private var progress = 1
        set(value) {
            val angle = (360 * (value / 100.0f)).toInt()
            if (value in 0..100) {
                text = Integer.toString(value) + "%"
                if (animator != null) {
                    animator?.cancel()
                    animator?.setIntValues(progress, angle)
                } else {
                    animator = ValueAnimator.ofInt(progress, angle)
                    animator?.interpolator = AccelerateInterpolator()
                    animator?.duration = 800
                    animator?.addUpdateListener { animation ->
                        field = animation.animatedValue as Int
                        invalidate()
                    }
                }
                animator?.start()
            }
        }

    private var text = ""

    private var strokeWidth: Int = 0

    init {
        attrs?.let {
            val attributes = context.theme.obtainStyledAttributes(attrs, R.styleable.PercentageCircle, 0, 0)
            strokeWidth = attributes.getDimensionPixelSize(R.styleable.PercentageCircle_stroke, 3)
            attributes.recycle()
        }

        paint = Paint(Paint.ANTI_ALIAS_FLAG)
        paint.style = Paint.Style.FILL
        paint.color = ContextCompat.getColor(getContext(), R.color.colorVividBlue)

        paintGray = Paint(Paint.ANTI_ALIAS_FLAG)
        paint.style = Paint.Style.FILL
        paintGray.color = ContextCompat.getColor(getContext(), R.color.colorLightGray)

        centerPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        centerPaint.style = Paint.Style.FILL
        centerPaint.color = ContextCompat.getColor(getContext(), android.R.color.white)

        paintText = Paint(Paint.ANTI_ALIAS_FLAG)
        paintText.style = Paint.Style.FILL
        paintText.color = ContextCompat.getColor(getContext(), R.color.colorVividBlue)

        resize()
    }

    private fun resize() {
        val heightOrWidth = Math.min(width, height) //ALSO diameter
        strokeWidth = heightOrWidth / 15
        val top = strokeWidth / 2
        val bottom = heightOrWidth - strokeWidth / 2
        paint.strokeWidth = strokeWidth.toFloat()
        bounds = RectF(top.toFloat(), top.toFloat(), bottom.toFloat(), bottom.toFloat())
        boundsGray = RectF(top.toFloat(), top.toFloat(), bottom.toFloat(), bottom.toFloat())
        val newTop = top + strokeWidth
        val newBottom = bottom - strokeWidth
        centerBounds = RectF(newTop.toFloat(), newTop.toFloat(), newBottom.toFloat(), newBottom.toFloat())
        paintText.textSize = (heightOrWidth / 4).toFloat()
        invalidate()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        setMeasuredDimension(widthMeasureSpec, heightMeasureSpec)
        resize()
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        canvas.drawArc(boundsGray, 0f, 360f, true, paintGray)
        canvas.drawArc(bounds, -91f, progress.toFloat(), true, paint)
        canvas.drawArc(centerBounds, 0f, 360f, true, centerPaint)
        val xPos = canvas.width / 2
        val yPos = xPos - ((paintText.descent() + paintText.ascent()) / 2).toInt()
        canvas.drawText(text, (xPos - paintText.measureText(text).toInt() / 2).toFloat(), yPos.toFloat(), paintText)
    }
}